# Data Wrangling in Pandas
<img src='images/img_one.jpg' align=right>
The Journal of Data Science defines "data science" as almost everything that has something to do with data. In a job, this translates to using data to have an impact on the organization by adding value. Most commonly it is to use and apply the data to solve complex business problems. One of the most common steps taken in data science work is data wrangling. The following is a concise guide on how to go about exploring, manipulating and reshaping data in python using the <a href="https://pandas.pydata.org/pandas-docs/stable/">pandas</a> library.
 
We will explore a breast cancer data set (credits: <a href="https://archive.ics.uci.edu/ml/datasets/breast+cancer+wisconsin+(original)">UCI</a>) and use pandas to clean, reshape, massage and give us a clean data set, all of this will help dramatically increase the quality of our data. Note: Data quality is KEY for optimal performance with machine learning algorithms.

The following pandas functionalities will be covered:
1. Data exploration — columns, unique values in a column, describe, duplicates
2. Dealing with missing values — quantifying missing values per column, filling & dropping missing values
3. Reshaping data — one hot encoding, pivot tables, joins, grouping and aggregating
4. Filtering data
5. Other — Making descriptive columns, element-wise conditional operations



## Data Exploration
Let us begin by reading in our dataset (csv file) into pandas and displaying the column names along with their data types. Also take a moment to view the entire dataset.

In [2]:
import pandas as pd
pd.options.display.max_columns =None
pd.options.display.max_rows =40
filename = 'data/breast_cancer_data.csv'
df = pd.read_csv(filename)

In [3]:
#start of by actually looking at your data set
df

patient_id  clump_thickness  cell_size_uniformity  cell_shape_uniformity  \
0       1000025              5.0                   1.0                      1   
1       1002945              5.0                   4.0                      4   
2       1015425              3.0                   1.0                      1   
3       1016277              6.0                   8.0                      8   
4       1017023              4.0                   1.0                      1   
5       1017122              8.0                  10.0                     10   
6       1018099              1.0                   NaN                      1   
7       1018561              2.0                   1.0                      2   
8       1033078              2.0                   1.0                      1   
9       1033078              4.0                   2.0                      1   
10      1035283              1.0                   1.0                      1   
11      1036172              2.0                   1.0                      1   
12      1041801              NaN                   3.0                      3   
13      1043999              1.0                   1.0                      1   
14      1044572              8.0                   7.0                      5   
15      1047630              7.0                   4.0                      6   
16      1048672              4.0                   1.0                      1   
17      1049815              4.0                   1.0                      1   
18      1050670             10.0                   7.0                      7   
19      1050718              6.0                   1.0                      1   
..          ...              ...                   ...                    ...   
679     1368882              2.0                   1.0                      1   
680     1369821             10.0                  10.0                     10   
681     1371026              5.0                  10.0                     10   
682     1371920              5.0                   1.0                      1   
683      466906              1.0                   1.0                      1   
684      466906              1.0                   1.0                      1   
685      534555              1.0                   1.0                      1   
686      536708              1.0                   1.0                      1   
687      566346              3.0                   1.0                      1   
688      603148              4.0                   1.0                      1   
689      654546              1.0                   1.0                      1   
690      654546              1.0                   1.0                      1   
691      695091              5.0                  10.0                     10   
692      714039              3.0                   1.0                      1   
693      763235              3.0                   1.0                      1   
694      776715              3.0                   1.0                      1   
695      841769              2.0                   1.0                      1   
696      888820              5.0                  10.0                     10   
697      897471              4.0                   8.0                      6   
698      897471              4.0                   8.0                      8   

     marginal_adhesion  single_ep_cell_size bare_nuclei  bland_chromatin  \
0                    1                    2           1              3.0   
1                    5                    7          10              3.0   
2                    1                    2           2              3.0   
3                    1                    3           4              3.0   
4                    3                    2           1              3.0   
5                    8                    7          10              9.0   
6                    1                    2          10              3

In [4]:
# What is the size of our dataset?
df.shape

(699, 12)

In [5]:
# Over here we see the columns names and their data types
df.dtypes

patient_id                 int64
clump_thickness          float64
cell_size_uniformity     float64
cell_shape_uniformity      int64
marginal_adhesion          int64
single_ep_cell_size        int64
bare_nuclei               object
bland_chromatin          float64
normal_nucleoli          float64
mitoses                    int64
class                     object
doctor_name               object
dtype: object

In the data we have the following columns as described by the source — Patient ID: id number, Clump Thickness: 1–10, Uniformity of Cell Size: 1–10, Uniformity of Cell Shape: 1–10, Marginal Adhesion: 1–10, Single Epithelial Cell Size: 1–10, Bare Nuclei: 1–10, Bland Chromatin: 1–10, Normal Nucleoli: 1–10, Mitoses: 1–10, Class: malignant or benign, Doctor name: 4 different doctors.

Based on this, we can assume that patient_id is a unique identifier, class is going to tell us whether the tumor is malignant (cancerous) or benign (not cancerous). The remaining columns are numeric medical descriptions of the tumor, except for the doctor_name which is a categorical feature.

Things to keep in mind — If our goal is to predict wether a tumor is cancerous or not based on the remaining features, we will have to one hot encode the categorical data and clean up the numerical data.
From our first output we see that bare_nuclei was read as an object data type although the description is numeric. Therefore we will need to change this.

To verify that our data matches up with the source we can use the describe option in pandas:

In [6]:
# This provides some statistics on the numerical data
df.describe()

patient_id  clump_thickness  cell_size_uniformity  \
count  6.990000e+02       698.000000            698.000000   
mean   1.071704e+06         4.416905              3.137536   
std    6.170957e+05         2.817673              3.052575   
min    6.163400e+04         1.000000              1.000000   
25%    8.706885e+05         2.000000              1.000000   
50%    1.171710e+06         4.000000              1.000000   
75%    1.238298e+06         6.000000              5.000000   
max    1.345435e+07        10.000000             10.000000   

       cell_shape_uniformity  marginal_adhesion  single_ep_cell_size  \
count             699.000000         699.000000           699.000000   
mean                3.207439           2.793991             3.216023   
std                 2.971913           2.843163             2.214300   
min                 1.000000           1.000000             1.000000   
25%                 1.000000           1.000000             2.000000   
50%                 1.000000           1.000000             2.000000   
75%                 5.000000           3.500000             4.000000   
max                10.000000          10.000000            10.000000   

       bland_chromatin  normal_nucleoli     mitoses  
count       695.000000       698.000000  699.000000  
mean          3.447482         2.868195    1.589413  
std           2.441191         3.055647    1.715078  
min           1.000000         1.000000    1.000000  
25%           2.000000         1.000000    1.000000  
50%           3.000000         1.000000    1.000000  
75%           5.000000         4.000000    1.000000  
max          10.000000        10.000000   10.000000

This neatly summarizes some statistical data for all numerical columns. For categorical data we can hand this by grouping together values:

In [7]:
# This aggreates the data by its column names, then we pass the aggregation function (size = count)
df.groupby(by =['class', 'doctor_name']).size()

class      doctor_name
benign     Dr. Doe        127
           Dr. Lee        121
           Dr. Smith      102
           Dr. Wong       108
malignant  Dr. Doe         58
           Dr. Lee         60
           Dr. Smith       74
           Dr. Wong        49
dtype: int64

## Dealing with missing values

With every dataset it is vital to evaluate the missing values. How many are there? Is it an error? Are there too many missing values? Does a missing value have a meaning relative to its context?
We can sum up the total missing values using the following:

In [8]:
#Dealing with missing values? How many np.nan per column?

df.isna().sum() 

AttributeError: 'DataFrame' object has no attribute 'isna'

Now that we have identified our missing values, we have a few options. We can fill them in with a certain value (zero, mean/max/median by column, string) or drop them by row. Since there are few missing values, we can drop the rows to avoid skewing the data in further analysis.

In [9]:
# # fill with zero
# df = df.fillna(0) 

In [10]:
df = df.dropna(axis = 0, how = 'any')  #drop rows with any column having np.nan values

#Rename columns
#df.rename(index =str, columns = {'patient_id':'patient_id'})
df

patient_id  clump_thickness  cell_size_uniformity  cell_shape_uniformity  \
0       1000025              5.0                   1.0                      1   
1       1002945              5.0                   4.0                      4   
2       1015425              3.0                   1.0                      1   
3       1016277              6.0                   8.0                      8   
4       1017023              4.0                   1.0                      1   
5       1017122              8.0                  10.0                     10   
7       1018561              2.0                   1.0                      2   
8       1033078              2.0                   1.0                      1   
9       1033078              4.0                   2.0                      1   
10      1035283              1.0                   1.0                      1   
11      1036172              2.0                   1.0                      1   
13      1043999              1.0                   1.0                      1   
14      1044572              8.0                   7.0                      5   
15      1047630              7.0                   4.0                      6   
16      1048672              4.0                   1.0                      1   
17      1049815              4.0                   1.0                      1   
18      1050670             10.0                   7.0                      7   
19      1050718              6.0                   1.0                      1   
20      1054590              7.0                   3.0                      2   
21      1054593             10.0                   5.0                      5   
..          ...              ...                   ...                    ...   
679     1368882              2.0                   1.0                      1   
680     1369821             10.0                  10.0                     10   
681     1371026              5.0                  10.0                     10   
682     1371920              5.0                   1.0                      1   
683      466906              1.0                   1.0                      1   
684      466906              1.0                   1.0                      1   
685      534555              1.0                   1.0                      1   
686      536708              1.0                   1.0                      1   
687      566346              3.0                   1.0                      1   
688      603148              4.0                   1.0                      1   
689      654546              1.0                   1.0                      1   
690      654546              1.0                   1.0                      1   
691      695091              5.0                  10.0                     10   
692      714039              3.0                   1.0                      1   
693      763235              3.0                   1.0                      1   
694      776715              3.0                   1.0                      1   
695      841769              2.0                   1.0                      1   
696      888820              5.0                  10.0                     10   
697      897471              4.0                   8.0                      6   
698      897471              4.0                   8.0                      8   

     marginal_adhesion  single_ep_cell_size bare_nuclei  bland_chromatin  \
0                    1                    2           1              3.0   
1                    5                    7          10              3.0   
2                    1                    2           2              3.0   
3                    1                    3           4              3.0   
4                    3                    2           1              3.0   
5                    8                    7          10              9.0   
7                    1                    2           1              3

This allows us to drop rows with any missing values in them.

## Inspecting duplicates
To view repeating rows we can start off by looking at the number of unique values in each column.

In [11]:
# Here we list all columns
df.columns

Index([u'patient_id', u'clump_thickness', u'cell_size_uniformity',
       u'cell_shape_uniformity', u'marginal_adhesion', u'single_ep_cell_size',
       u'bare_nuclei', u'bland_chromatin', u'normal_nucleoli', u'mitoses',
       u'class', u'doctor_name'],
      dtype='object')

In [12]:
len(df)

690

In [13]:
# Its good to inspect your unique key identifier
df.nunique()

patient_id               637
clump_thickness           10
cell_size_uniformity      10
cell_shape_uniformity     10
marginal_adhesion         10
single_ep_cell_size       10
bare_nuclei               11
bland_chromatin           10
normal_nucleoli           10
mitoses                    9
class                      2
doctor_name                4
dtype: int64

We see here that although there are 690 rows, there are only 637 unique patient_id’s. This could mean that some patient appear more than once in the dataset. To isolate these patients and view their data, we use the following:

In [14]:
# This shows rows that show up more than once and have the exact same column values. 
df[df.duplicated(keep = 'last')]

patient_id  clump_thickness  cell_size_uniformity  cell_shape_uniformity  \
168     1198641              3.0                   1.0                      1   

     marginal_adhesion  single_ep_cell_size bare_nuclei  bland_chromatin  \
168                  1                    2           1              3.0   

     normal_nucleoli  mitoses   class doctor_name  
168              1.0        1  benign     Dr. Lee

In [15]:
# # This shows all instances where patient_id shows up more than once, but may have varying column values
df[df.duplicated(subset = 'patient_id', keep =False)].sort_values('patient_id')

patient_id  clump_thickness  cell_size_uniformity  cell_shape_uniformity  \
267      320675              3.0                   3.0                      5   
272      320675              3.0                   3.0                      5   
575      385103              5.0                   1.0                      2   
269      385103              1.0                   1.0                      1   
271      411453              5.0                   1.0                      1   
607      411453              1.0                   1.0                      1   
684      466906              1.0                   1.0                      1   
683      466906              1.0                   1.0                      1   
371      493452              1.0                   1.0                      3   
372      493452              4.0                   1.0                      2   
290      560680              1.0                   1.0                      1   
374      560680              3.0                   1.0                      2   
689      654546              1.0                   1.0                      1   
690      654546              1.0                   1.0                      1   
577      695091              1.0                   1.0                      1   
691      695091              5.0                  10.0                     10   
338      704097              1.0                   1.0                      1   
314      704097              1.0                   1.0                      1   
322      733639              3.0                   1.0                      1   
321      733639              3.0                   1.0                      1   
..          ...              ...                   ...                    ...   
430     1276091              1.0                   3.0                      1   
431     1276091              5.0                   1.0                      1   
241     1276091              3.0                   1.0                      1   
462     1276091              6.0                   1.0                      1   
639     1277792              5.0                   1.0                      1   
638     1277792              4.0                   1.0                      1   
434     1293439              6.0                   9.0                      7   
433     1293439              3.0                   2.0                      2   
644     1299596              2.0                   1.0                      1   
467     1299596              6.0                   6.0                      6   
552     1299924              3.0                   2.0                      2   
511     1299924              5.0                   1.0                      1   
517     1320077              1.0                   1.0                      1   
516     1320077              1.0                   1.0                      1   
561     1321942              5.0                   1.0                      1   
560     1321942              5.0                   1.0                      1   
660     1339781              1.0                   1.0                      1   
661     1339781              4.0                   1.0                      1   
672     1354840              2.0                   1.0                      1   
673     1354840              5.0                   3.0                      2   

     marginal_adhesion  single_ep_cell_size bare_nuclei  bland_chromatin  \
267                  2                    3          10              7.0   
272                  2                    3          10              7.0   
575                  1                    2           1              3.0   
269                  1                    2           1              3.0   
271                  1                    2           1              3.0   
607                  1                    2           1              1.0   
684                  1                    2           1              1

In [16]:
#Now that I have seen that there are some duplicates, I am going to go ahead and remove any duplicate rows

#df = df.drop_duplicates(subset = None, keep ='first')

In [17]:
repeat_patients = df.groupby(by = 'patient_id').size().sort_values(ascending =False)
repeat_patients

patient_id
1182404    6
1276091    5
1198641    3
1033078    2
1238777    2
1143978    2
466906     2
1321942    2
493452     2
1320077    2
560680     2
1116192    2
1116116    2
1115293    2
1299924    2
1299596    2
1114570    2
654546     2
1293439    2
1105524    2
          ..
1208301    1
1211202    1
1217952    1
1211265    1
1217717    1
1217264    1
1217051    1
1216947    1
1216694    1
1214966    1
1214556    1
1214092    1
1213784    1
1213383    1
1213375    1
1213273    1
1212251    1
1212232    1
1211594    1
61634      1
Length: 637, dtype: int64

This shows that one patient shows up in the data 6 times!

## Filtering data
If we want to remove patients that show up more that 2 times in the data set.

In [18]:
filtered_patients = repeat_patients[repeat_patients > 2].to_frame().reset_index()
filtered_df = df[~df.patient_id.isin(filtered_patients.patient_id)]
filtered_df

patient_id  clump_thickness  cell_size_uniformity  cell_shape_uniformity  \
0       1000025              5.0                   1.0                      1   
1       1002945              5.0                   4.0                      4   
2       1015425              3.0                   1.0                      1   
3       1016277              6.0                   8.0                      8   
4       1017023              4.0                   1.0                      1   
5       1017122              8.0                  10.0                     10   
7       1018561              2.0                   1.0                      2   
8       1033078              2.0                   1.0                      1   
9       1033078              4.0                   2.0                      1   
10      1035283              1.0                   1.0                      1   
11      1036172              2.0                   1.0                      1   
13      1043999              1.0                   1.0                      1   
14      1044572              8.0                   7.0                      5   
15      1047630              7.0                   4.0                      6   
16      1048672              4.0                   1.0                      1   
17      1049815              4.0                   1.0                      1   
18      1050670             10.0                   7.0                      7   
19      1050718              6.0                   1.0                      1   
20      1054590              7.0                   3.0                      2   
21      1054593             10.0                   5.0                      5   
..          ...              ...                   ...                    ...   
679     1368882              2.0                   1.0                      1   
680     1369821             10.0                  10.0                     10   
681     1371026              5.0                  10.0                     10   
682     1371920              5.0                   1.0                      1   
683      466906              1.0                   1.0                      1   
684      466906              1.0                   1.0                      1   
685      534555              1.0                   1.0                      1   
686      536708              1.0                   1.0                      1   
687      566346              3.0                   1.0                      1   
688      603148              4.0                   1.0                      1   
689      654546              1.0                   1.0                      1   
690      654546              1.0                   1.0                      1   
691      695091              5.0                  10.0                     10   
692      714039              3.0                   1.0                      1   
693      763235              3.0                   1.0                      1   
694      776715              3.0                   1.0                      1   
695      841769              2.0                   1.0                      1   
696      888820              5.0                  10.0                     10   
697      897471              4.0                   8.0                      6   
698      897471              4.0                   8.0                      8   

     marginal_adhesion  single_ep_cell_size bare_nuclei  bland_chromatin  \
0                    1                    2           1              3.0   
1                    5                    7          10              3.0   
2                    1                    2           2              3.0   
3                    1                    3           4              3.0   
4                    3                    2           1              3.0   
5                    8                    7          10              9.0   
7                    1                    2           1              3

If we did not have the tilde (“~”) we would get all individuals that repeat more than twice. By adding a tilde the pandas boolean series is reversed and thus the resulting data frame is of those that do NOT repeat more than twice.

## Reshaping data
The dataset has elements of categorical data in the “doctor_name” column. To feed this data into a machine learning pipeline, we will need to convert it into a one hot encoded column. This can be done with a sci-kit learn package, however we will do it in pandas to demonstrate the pivoting and merging functionality. Start off by creating a new dataframe with the categorical data.

In [19]:
categorical_df = df[['patient_id','doctor_name']]
categorical_df

patient_id doctor_name
0       1000025     Dr. Doe
1       1002945   Dr. Smith
2       1015425     Dr. Lee
3       1016277   Dr. Smith
4       1017023    Dr. Wong
5       1017122   Dr. Smith
7       1018561   Dr. Smith
8       1033078   Dr. Smith
9       1033078     Dr. Doe
10      1035283     Dr. Doe
11      1036172   Dr. Smith
13      1043999    Dr. Wong
14      1044572     Dr. Doe
15      1047630     Dr. Lee
16      1048672     Dr. Lee
17      1049815   Dr. Smith
18      1050670   Dr. Smith
19      1050718    Dr. Wong
20      1054590     Dr. Doe
21      1054593   Dr. Smith
..          ...         ...
679     1368882     Dr. Doe
680     1369821     Dr. Lee
681     1371026    Dr. Wong
682     1371920    Dr. Wong
683      466906     Dr. Lee
684      466906    Dr. Wong
685      534555     Dr. Doe
686      536708     Dr. Doe
687      566346     Dr. Lee
688      603148   Dr. Smith
689      654546     Dr. Lee
690      654546     Dr. Doe
691      695091    Dr. Wong
692      714039    Dr. Wong
693      763235     Dr. Lee
694      776715     Dr. Lee
695      841769   Dr. Smith
696      888820     Dr. Lee
697      897471     Dr. Lee
698      897471    Dr. Wong

[690 rows x 2 columns]

In [20]:
# This specifies all rows (':') and column name 'doctor_count'
categorical_df.loc[:,'doctor_count'] = 1

/Users/teddyrosenbaum/anaconda/lib/python2.7/site-packages/pandas/core/indexing.py:337: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  self.obj[key] = _infer_fill_value(value)
/Users/teddyrosenbaum/anaconda/lib/python2.7/site-packages/pandas/core/indexing.py:517: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  self.obj[item] = s


In [21]:
categorical_df

patient_id doctor_name  doctor_count
0       1000025     Dr. Doe             1
1       1002945   Dr. Smith             1
2       1015425     Dr. Lee             1
3       1016277   Dr. Smith             1
4       1017023    Dr. Wong             1
5       1017122   Dr. Smith             1
7       1018561   Dr. Smith             1
8       1033078   Dr. Smith             1
9       1033078     Dr. Doe             1
10      1035283     Dr. Doe             1
11      1036172   Dr. Smith             1
13      1043999    Dr. Wong             1
14      1044572     Dr. Doe             1
15      1047630     Dr. Lee             1
16      1048672     Dr. Lee             1
17      1049815   Dr. Smith             1
18      1050670   Dr. Smith             1
19      1050718    Dr. Wong             1
20      1054590     Dr. Doe             1
21      1054593   Dr. Smith             1
..          ...         ...           ...
679     1368882     Dr. Doe             1
680     1369821     Dr. Lee             1
681     1371026    Dr. Wong             1
682     1371920    Dr. Wong             1
683      466906     Dr. Lee             1
684      466906    Dr. Wong             1
685      534555     Dr. Doe             1
686      536708     Dr. Doe             1
687      566346     Dr. Lee             1
688      603148   Dr. Smith             1
689      654546     Dr. Lee             1
690      654546     Dr. Doe             1
691      695091    Dr. Wong             1
692      714039    Dr. Wong             1
693      763235     Dr. Lee             1
694      776715     Dr. Lee             1
695      841769   Dr. Smith             1
696      888820     Dr. Lee             1
697      897471     Dr. Lee             1
698      897471    Dr. Wong             1

[690 rows x 3 columns]

We add a column an extra column to identify which doctor a patient deals with. Pivot this table so that we only have numerical values in the cells and the columns become the doctors’ name. Then fill in the empty cells with 0.

In [22]:
doctors_one_hot_encoded = pd.pivot_table( categorical_df,
                                  index = categorical_df.index, 
                                  columns = ['doctor_name'], 
                                   values = ['doctor_count'] )
doctors_one_hot_encoded = doctors_one_hot_encoded.fillna(0)
doctors_one_hot_encoded

doctor_count                           
doctor_name      Dr. Doe Dr. Lee Dr. Smith Dr. Wong
0                    1.0     0.0       0.0      0.0
1                    0.0     0.0       1.0      0.0
2                    0.0     1.0       0.0      0.0
3                    0.0     0.0       1.0      0.0
4                    0.0     0.0       0.0      1.0
5                    0.0     0.0       1.0      0.0
7                    0.0     0.0       1.0      0.0
8                    0.0     0.0       1.0      0.0
9                    1.0     0.0       0.0      0.0
10                   1.0     0.0       0.0      0.0
11                   0.0     0.0       1.0      0.0
13                   0.0     0.0       0.0      1.0
14                   1.0     0.0       0.0      0.0
15                   0.0     1.0       0.0      0.0
16                   0.0     1.0       0.0      0.0
17                   0.0     0.0       1.0      0.0
18                   0.0     0.0       1.0      0.0
19                   0.0     0.0       0.0      1.0
20                   1.0     0.0       0.0      0.0
21                   0.0     0.0       1.0      0.0
..                   ...     ...       ...      ...
679                  1.0     0.0       0.0      0.0
680                  0.0     1.0       0.0      0.0
681                  0.0     0.0       0.0      1.0
682                  0.0     0.0       0.0      1.0
683                  0.0     1.0       0.0      0.0
684                  0.0     0.0       0.0      1.0
685                  1.0     0.0       0.0      0.0
686                  1.0     0.0       0.0      0.0
687                  0.0     1.0       0.0      0.0
688                  0.0     0.0       1.0      0.0
689                  0.0     1.0       0.0      0.0
690                  1.0     0.0       0.0      0.0
691                  0.0     0.0       0.0      1.0
692                  0.0     0.0       0.0      1.0
693                  0.0     1.0       0.0      0.0
694                  0.0     1.0       0.0      0.0
695                  0.0     0.0       1.0      0.0
696                  0.0     1.0       0.0      0.0
697                  0.0     1.0       0.0      0.0
698                  0.0     0.0       0.0      1.0

[690 rows x 4 columns]

Then drop the multiIndex columns:


In [23]:
doctors_one_hot_encoded.columns = doctors_one_hot_encoded.columns.droplevel()
doctors_one_hot_encoded

doctor_name  Dr. Doe  Dr. Lee  Dr. Smith  Dr. Wong
0                1.0      0.0        0.0       0.0
1                0.0      0.0        1.0       0.0
2                0.0      1.0        0.0       0.0
3                0.0      0.0        1.0       0.0
4                0.0      0.0        0.0       1.0
5                0.0      0.0        1.0       0.0
7                0.0      0.0        1.0       0.0
8                0.0      0.0        1.0       0.0
9                1.0      0.0        0.0       0.0
10               1.0      0.0        0.0       0.0
11               0.0      0.0        1.0       0.0
13               0.0      0.0        0.0       1.0
14               1.0      0.0        0.0       0.0
15               0.0      1.0        0.0       0.0
16               0.0      1.0        0.0       0.0
17               0.0      0.0        1.0       0.0
18               0.0      0.0        1.0       0.0
19               0.0      0.0        0.0       1.0
20               1.0      0.0        0.0       0.0
21               0.0      0.0        1.0       0.0
..               ...      ...        ...       ...
679              1.0      0.0        0.0       0.0
680              0.0      1.0        0.0       0.0
681              0.0      0.0        0.0       1.0
682              0.0      0.0        0.0       1.0
683              0.0      1.0        0.0       0.0
684              0.0      0.0        0.0       1.0
685              1.0      0.0        0.0       0.0
686              1.0      0.0        0.0       0.0
687              0.0      1.0        0.0       0.0
688              0.0      0.0        1.0       0.0
689              0.0      1.0        0.0       0.0
690              1.0      0.0        0.0       0.0
691              0.0      0.0        0.0       1.0
692              0.0      0.0        0.0       1.0
693              0.0      1.0        0.0       0.0
694              0.0      1.0        0.0       0.0
695              0.0      0.0        1.0       0.0
696              0.0      1.0        0.0       0.0
697              0.0      1.0        0.0       0.0
698              0.0      0.0        0.0       1.0

[690 rows x 4 columns]

We can now join this back to our main table. Typically a left join in pandas looks like this:

`leftJoin_df = pd.merge(df1, df2, on ='col_name', how='left')`

However we are joining on the index so we pass the “left_index” and “right_index” option to specify that the join key is the index of both tables



In [24]:
combined_df = pd.merge(df, doctors_one_hot_encoded, left_index = True,right_index =True, how ='left')
combined_df

patient_id  clump_thickness  cell_size_uniformity  cell_shape_uniformity  \
0       1000025              5.0                   1.0                      1   
1       1002945              5.0                   4.0                      4   
2       1015425              3.0                   1.0                      1   
3       1016277              6.0                   8.0                      8   
4       1017023              4.0                   1.0                      1   
5       1017122              8.0                  10.0                     10   
7       1018561              2.0                   1.0                      2   
8       1033078              2.0                   1.0                      1   
9       1033078              4.0                   2.0                      1   
10      1035283              1.0                   1.0                      1   
11      1036172              2.0                   1.0                      1   
13      1043999              1.0                   1.0                      1   
14      1044572              8.0                   7.0                      5   
15      1047630              7.0                   4.0                      6   
16      1048672              4.0                   1.0                      1   
17      1049815              4.0                   1.0                      1   
18      1050670             10.0                   7.0                      7   
19      1050718              6.0                   1.0                      1   
20      1054590              7.0                   3.0                      2   
21      1054593             10.0                   5.0                      5   
..          ...              ...                   ...                    ...   
679     1368882              2.0                   1.0                      1   
680     1369821             10.0                  10.0                     10   
681     1371026              5.0                  10.0                     10   
682     1371920              5.0                   1.0                      1   
683      466906              1.0                   1.0                      1   
684      466906              1.0                   1.0                      1   
685      534555              1.0                   1.0                      1   
686      536708              1.0                   1.0                      1   
687      566346              3.0                   1.0                      1   
688      603148              4.0                   1.0                      1   
689      654546              1.0                   1.0                      1   
690      654546              1.0                   1.0                      1   
691      695091              5.0                  10.0                     10   
692      714039              3.0                   1.0                      1   
693      763235              3.0                   1.0                      1   
694      776715              3.0                   1.0                      1   
695      841769              2.0                   1.0                      1   
696      888820              5.0                  10.0                     10   
697      897471              4.0                   8.0                      6   
698      897471              4.0                   8.0                      8   

     marginal_adhesion  single_ep_cell_size bare_nuclei  bland_chromatin  \
0                    1                    2           1              3.0   
1                    5                    7          10              3.0   
2                    1                    2           2              3.0   
3                    1                    3           4              3.0   
4                    3                    2           1              3.0   
5                    8                    7          10              9.0   
7                    1                    2           1              3

We can drop the column that we no longer need by the following:

In [25]:
combined_df = combined_df.drop(columns=['doctor_name'])
combined_df

TypeError: drop() got an unexpected keyword argument 'columns'

## Row-wise Operations
Another key component in data wrangling is having the ability to conduct row-wise or column wise operations. Examples of this are; rename elements within a column based on its value and create a new column that yields a specific value based on multiple attributes within the row.

For this example lets create a new column that categorizes a patients cell as normal or abnormal based on its attributes. We first define our function and the operation that it will be doing.

In [26]:
def celltypelabel(x):
    if ((x['cell_size_uniformity'] < 5) &      (x['cell_shape_uniformity'] < 5)):
        
        return('normal')
    else:
        return('abnormal')

Then we use the pandas apply function to run the celltypelabel(x) function on the dataframe.

In [27]:
combined_df['cell_type_label'] = combined_df.apply(lambda x: celltypelabel(x), axis=1)

In [28]:
combined_df

patient_id  clump_thickness  cell_size_uniformity  cell_shape_uniformity  \
0       1000025              5.0                   1.0                      1   
1       1002945              5.0                   4.0                      4   
2       1015425              3.0                   1.0                      1   
3       1016277              6.0                   8.0                      8   
4       1017023              4.0                   1.0                      1   
5       1017122              8.0                  10.0                     10   
7       1018561              2.0                   1.0                      2   
8       1033078              2.0                   1.0                      1   
9       1033078              4.0                   2.0                      1   
10      1035283              1.0                   1.0                      1   
11      1036172              2.0                   1.0                      1   
13      1043999              1.0                   1.0                      1   
14      1044572              8.0                   7.0                      5   
15      1047630              7.0                   4.0                      6   
16      1048672              4.0                   1.0                      1   
17      1049815              4.0                   1.0                      1   
18      1050670             10.0                   7.0                      7   
19      1050718              6.0                   1.0                      1   
20      1054590              7.0                   3.0                      2   
21      1054593             10.0                   5.0                      5   
..          ...              ...                   ...                    ...   
679     1368882              2.0                   1.0                      1   
680     1369821             10.0                  10.0                     10   
681     1371026              5.0                  10.0                     10   
682     1371920              5.0                   1.0                      1   
683      466906              1.0                   1.0                      1   
684      466906              1.0                   1.0                      1   
685      534555              1.0                   1.0                      1   
686      536708              1.0                   1.0                      1   
687      566346              3.0                   1.0                      1   
688      603148              4.0                   1.0                      1   
689      654546              1.0                   1.0                      1   
690      654546              1.0                   1.0                      1   
691      695091              5.0                  10.0                     10   
692      714039              3.0                   1.0                      1   
693      763235              3.0                   1.0                      1   
694      776715              3.0                   1.0                      1   
695      841769              2.0                   1.0                      1   
696      888820              5.0                  10.0                     10   
697      897471              4.0                   8.0                      6   
698      897471              4.0                   8.0                      8   

     marginal_adhesion  single_ep_cell_size bare_nuclei  bland_chromatin  \
0                    1                    2           1              3.0   
1                    5                    7          10              3.0   
2                    1                    2           2              3.0   
3                    1                    3           4              3.0   
4                    3                    2           1              3.0   
5                    8                    7          10              9.0   
7                    1                    2           1              3

## Conclusion
Although some of these data manipulation steps can be done in SAS and excel. Doing it in python not only allows you to connect the data to vast open source resources in computer vision, machine and deep learning, but also for ETL automation purposes and more.

# Homework
1. Hot encode a new column in this dataset for cancerous (1) or not cancerous (0).

In [32]:
def cancerous(x):
    if (x['class'] == 'malignant'): 
        return 1
    else:
        return 0

In [33]:
combined_df['cancerous'] = combined_df.apply(lambda x: cancerous(x), axis=1)

In [34]:
combined_df

patient_id  clump_thickness  cell_size_uniformity  cell_shape_uniformity  \
0       1000025              5.0                   1.0                      1   
1       1002945              5.0                   4.0                      4   
2       1015425              3.0                   1.0                      1   
3       1016277              6.0                   8.0                      8   
4       1017023              4.0                   1.0                      1   
5       1017122              8.0                  10.0                     10   
7       1018561              2.0                   1.0                      2   
8       1033078              2.0                   1.0                      1   
9       1033078              4.0                   2.0                      1   
10      1035283              1.0                   1.0                      1   
11      1036172              2.0                   1.0                      1   
13      1043999              1.0                   1.0                      1   
14      1044572              8.0                   7.0                      5   
15      1047630              7.0                   4.0                      6   
16      1048672              4.0                   1.0                      1   
17      1049815              4.0                   1.0                      1   
18      1050670             10.0                   7.0                      7   
19      1050718              6.0                   1.0                      1   
20      1054590              7.0                   3.0                      2   
21      1054593             10.0                   5.0                      5   
..          ...              ...                   ...                    ...   
679     1368882              2.0                   1.0                      1   
680     1369821             10.0                  10.0                     10   
681     1371026              5.0                  10.0                     10   
682     1371920              5.0                   1.0                      1   
683      466906              1.0                   1.0                      1   
684      466906              1.0                   1.0                      1   
685      534555              1.0                   1.0                      1   
686      536708              1.0                   1.0                      1   
687      566346              3.0                   1.0                      1   
688      603148              4.0                   1.0                      1   
689      654546              1.0                   1.0                      1   
690      654546              1.0                   1.0                      1   
691      695091              5.0                  10.0                     10   
692      714039              3.0                   1.0                      1   
693      763235              3.0                   1.0                      1   
694      776715              3.0                   1.0                      1   
695      841769              2.0                   1.0                      1   
696      888820              5.0                  10.0                     10   
697      897471              4.0                   8.0                      6   
698      897471              4.0                   8.0                      8   

     marginal_adhesion  single_ep_cell_size bare_nuclei  bland_chromatin  \
0                    1                    2           1              3.0   
1                    5                    7          10              3.0   
2                    1                    2           2              3.0   
3                    1                    3           4              3.0   
4                    3                    2           1              3.0   
5                    8                    7          10              9.0   
7                    1                    2           1              3